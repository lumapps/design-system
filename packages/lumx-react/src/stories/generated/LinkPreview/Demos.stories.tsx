/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/link-preview/Demos' };

export { App as Big } from './big';
export { App as Default } from './default';
