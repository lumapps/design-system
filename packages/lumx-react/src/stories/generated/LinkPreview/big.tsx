/home/gcornut/git/design-system/packages/site-demo/content/product/components/link-preview/react/big.tsx