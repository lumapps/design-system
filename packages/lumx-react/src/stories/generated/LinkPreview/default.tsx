../../../../../site-demo/content/product/components/link-preview/react/default.tsx