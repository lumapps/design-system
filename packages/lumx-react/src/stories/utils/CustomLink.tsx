/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
