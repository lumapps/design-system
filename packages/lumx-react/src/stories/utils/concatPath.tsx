export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
