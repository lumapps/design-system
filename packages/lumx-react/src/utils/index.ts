/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
