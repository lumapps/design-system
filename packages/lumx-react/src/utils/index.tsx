export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
