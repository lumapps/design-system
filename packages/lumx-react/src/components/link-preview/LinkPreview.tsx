import React, { forwardRef, useCallback } from 'react';

import classNames from 'classnames';

import {
    AspectRatio,
    ColorPalette,
    ColorVariant,
    Link,
    LinkProps,
    Size,
    Theme,
    Thumbnail,
    ThumbnailProps,
} from '@lumx/react';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

/**
 * Defines the props of the component.
 */
export interface LinkPreviewProps extends GenericProps {
    /** Description (either a string, or sanitized html). */
    description?: string | { __html: string };
    /** Link URL. */
    link: string;
    /** Props to pass to the link (minus those already set by the LinkPreview props). */
    linkProps?: Omit<LinkProps, 'color' | 'colorVariant' | 'href' | 'target'>;
    /** Size variant. */
    size?: Extract<Size, 'regular' | 'big'>;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Thumbnail for the link preview. */
    thumbnailProps?: ThumbnailProps;
    /** Title. */
    title?: string;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'LinkPreview';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<LinkPreviewProps> = {
    size: Size.regular,
    theme: Theme.light,
};

/**
 * LinkPreview component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const LinkPreview: Comp<LinkPreviewProps, HTMLDivElement> = forwardRef((props, ref) => {
    const { className, description, link, linkProps, size, theme, thumbnailProps, title, ...forwardedProps } = props;

    //TODO: a11y
    const goToUrl = useCallback(() => window.open(link, '_blank'), [link]);

    return (
        <div
            ref={ref}
            {...forwardedProps}
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                    size: size === Size.big && thumbnailProps ? Size.big : Size.regular,
                    theme,
                }),
            )}
        >
            <div className={`${CLASSNAME}__wrapper`}>
                {thumbnailProps && (
                    <div className={`${CLASSNAME}__thumbnail`}>
                        <Thumbnail
                            {...thumbnailProps}
                            onClick={goToUrl}
                            role="link"
                            aspectRatio={AspectRatio.free}
                            fillHeight
                        />
                    </div>
                )}

                <div className={`${CLASSNAME}__container`}>
                    {title && (
                        <div className={`${CLASSNAME}__title`}>
                            <Link
                                {...linkProps}
                                target="_blank"
                                href={link}
                                color={theme === Theme.light ? ColorPalette.dark : ColorPalette.light}
                                colorVariant={ColorVariant.N}
                            >
                                {title}
                            </Link>
                        </div>
                    )}
                    {description && <p className={`${CLASSNAME}__description`}>{description}</p>}

                    <div className={`${CLASSNAME}__link`}>
                        <Link
                            {...linkProps}
                            className={classNames(`${CLASSNAME}__link`, linkProps?.className)}
                            target="_blank"
                            href={link}
                            color={theme === Theme.light ? ColorPalette.blue : ColorPalette.light}
                            colorVariant={ColorVariant.N}
                        >
                            {link}
                        </Link>
                    </div>
                </div>
            </div>
        </div>
    );
});

LinkPreview.displayName = COMPONENT_NAME;
LinkPreview.className = CLASSNAME;
LinkPreview.defaultProps = DEFAULT_PROPS;
