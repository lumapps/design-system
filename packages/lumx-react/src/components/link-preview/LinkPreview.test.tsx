import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';
import { Thumbnail } from '@lumx/react';

import { Size, Theme } from '..';
import { LinkPreview, LinkPreviewProps } from './LinkPreview';

const DEFAULT_PROPS = LinkPreview.defaultProps as any;
const CLASSNAME = LinkPreview.className as string;

type SetupProps = Partial<LinkPreviewProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}, shallowRendering = true) => {
    const props: any = { ...propsOverride };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper: Wrapper = renderer(<LinkPreview {...props} />);

    return {
        thumbnail: wrapper.find(Thumbnail),
        props,
        wrapper,
    };
};

describe(`<${LinkPreview.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });

        it('should render correctly', () => {
            const { wrapper } = setup({ size: Size.big });
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default props', () => {
            const { wrapper, thumbnail } = setup();
            ['size', 'theme'].forEach((prop: string) => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: DEFAULT_PROPS[prop] }),
                );
            });
            expect(thumbnail).not.toExist();
        });

        it('should pass className prop to the wrapper', () => {
            const expectedClassName = 'must-be-set';
            const { wrapper } = setup({ className: expectedClassName });

            expect(wrapper).toHaveClassName(expectedClassName);
        });

        it('should set --theme-dark class variant on wrapper if theme = Theme.dark', () => {
            const { wrapper } = setup({ theme: Theme.dark });

            expect(wrapper).toHaveClassName(getBasicClass({ prefix: CLASSNAME, type: 'theme', value: Theme.dark }));
        });
    });

    // 3. Test events.
    describe('Events', () => {
        const expectedUrl = 'https://expected.url';
        const { thumbnail } = setup({
            link: expectedUrl,
            thumbnailProps: { image: 'https://expected.url/image.png', alt: 'Alt' },
        });
        window.open = jest.fn();

        thumbnail.simulate('click');

        expect(window.open).toHaveBeenCalledWith(expectedUrl, '_blank');
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Nothing to do here.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
