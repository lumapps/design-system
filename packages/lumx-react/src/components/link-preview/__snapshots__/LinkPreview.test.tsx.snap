// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkPreview> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-link-preview lumx-link-preview--size-regular lumx-link-preview--theme-light"
>
  <div
    className="lumx-link-preview__wrapper"
  >
    <div
      className="lumx-link-preview__container"
    >
      <div
        className="lumx-link-preview__link"
      >
        <Link
          className="lumx-link-preview__link"
          color="primary"
          colorVariant="N"
          target="_blank"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<LinkPreview> Snapshots and structure should render correctly 2`] = `
<div
  className="lumx-link-preview lumx-link-preview--size-regular lumx-link-preview--theme-light"
>
  <div
    className="lumx-link-preview__wrapper"
  >
    <div
      className="lumx-link-preview__container"
    >
      <div
        className="lumx-link-preview__link"
      >
        <Link
          className="lumx-link-preview__link"
          color="primary"
          colorVariant="N"
          target="_blank"
        />
      </div>
    </div>
  </div>
</div>
`;
